### ISSUE TITLE
Cloudflare Load Balancer Pool deletes automatically added origins for Argo Tunnels

### ISSUE THREAD

**Issue by nin-ja-b** [cloudflare/terraform-provider-cloudflare#971]
### Terraform version
0.14.2

### Affected resource(s)
cloudflare_load_balancer_pool

### Terraform configuration files
```
resource "cloudflare_load_balancer_pool" "ingress" {
  name = local.lb_pool_name
  description = "Load balancer pool"
  enabled = true
  minimum_origins = 1
  monitor = cloudflare_load_balancer_monitor.monitor.id
  notification_email = "me@example.com"
  origins {
    address = "example.com"
    name    = "dummy-entry"
    enabled = false
  }
}
```
This is the first problem here. There is no way not to specify any origins so I need a dummy origin which isn't gread because in our plan we can only have 20. 

### Debug output
``` 
# cloudflare_load_balancer_pool.ingress will be updated in-place
  ~ resource "cloudflare_load_balancer_pool" "ingress" {
        id                 = "44444444444"
        name               = "argo-tunnel-ingress.example.com"
        # (8 unchanged attributes hidden)

      - origins {
          - address = "tunnel:72k858439574389579348..." -> null
          - enabled = true -> null
          - name    = "0-46325b-5f24-4f95-345436677-35435" -> null
          - weight  = 1 -> null
        } ..........
```

### Expected behavior
I am creating the load balancer pool with terraform. The argo tunnels for this Load Balancer Pool are created automatically by running the cloudflared docker image as a sidecar in my Kubernetes Ingress-Nginx and passing in the name of the Load Balancer Pool. 
After the argo tunnels are added I want to be able to run terraform again without terraform deleting the automatically created tunnels.  

### Actual behavior
Terraform wants to delete the automatically added argo tunnels from the loadbalancer Pool.

### Steps to reproduce
- terraform apply
- add tunnels via cloudflared image or UI
- terraform apply

### Community note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)
  to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull
  request, please leave a comment

**Comment by nin-ja-b** [comment]
Never mind, I'm just going to use:
```
  lifecycle {
    ignore_changes = [origins]
  }
```

**Comment by jacobbednarz** [comment]
Just so I understand your use case, are you trying to create and manage resources in Terraform _and_ external sources such as scripts or the dashboard? If that is the case, you're always going to be at odds with Terraform as the expectation is that it maintains "the golden state" and will always try to correct the current state to match it. 

Your workaround is the cleanest way I know of to keep both sources happy, however, if you are managing changes outside of Terraform, it may be best not to manage it within Terraform at all should problems persist.

**Comment by nin-ja-b** [comment]
Yes, you are a 100% right. There should be one truth and if there is a way I would love to go that way.
In the end what I want to do is expose my ingress-nginx running in Kubernetes via cloudflare argo tunnels and I want to do that without clicking things together in cloudflare.
So far we have used the cloudflared docker image to create those tunnels which are then automatically added to the load balancer pool. 
I'd be super happy to create the argo tunnel via Terraform too and that TF resource already exists. So I can create the argo tunnel, but so far I wasn't able to get my cloudflare sidecar to work with an already created argo tunnel. 
Maybe you got an idea? 
This is the sidecar container. It needs an existing loadbalancer pool and that damn cert.pem:
´´´
  extraContainers:
    - name: tunnel
      image: docker.io/cloudflare/cloudflared:2021.2.4
      imagePullPolicy: Always
      command: ["cloudflared", "tunnel"]
      args:
        - --url=http://127.0.0.1:80
        - --hostname=${url}
        - --lb-pool=${lb_pool}
        - --origincert=/etc/cloudflared/cert.pem
        - --no-autoupdate
      volumeMounts:
        - mountPath: /etc/cloudflared
          name: tunnel-secret
          readOnly: true
  extraVolumes:
    ## Additional volumes to the controller pod.
    - name: tunnel-secret
      secret:
        secretName: ${secret_name}
```

**Comment by jacobbednarz** [comment]
I can ask around internally but I don't think so. Load balancer pools don't make sense without origins so we can't really make them optional. Do you need them in Terraform if your origin is managing it in code?

**Comment by nin-ja-b** [comment]
I do, because I need to create the pool so that cloudflared can write the tunnels somewhere and via the UI would be even worse, but for this issue I think lifecycle - ignore_changes is fine. :woman_shrugging: 
You did bring me back to my first approach (that i didn't get to work before) though. It would be cleaner to create all resources in Terraform and to use the sidecar container only for connecting to the already existing tunnel.
So far I just wasn't able to get it running, but let me try again :)

**Comment by nin-ja-b** [comment]
So I am just going to post some code here, because it's so hard to find any of it anywhere. 
The tunnel is up and running, but unfortunately throwing errors still. I might need to go back to what I had before.

```
resource "kubernetes_secret" "cloudflare" {
  metadata {
    name      = local.tunnel_secret_name
    namespace = "ingress-nginx"
  }
  type = "Opaque"
  data = {
    "credentials.json" = jsonencode({
      "AccountTag"   = var.cloudflare_account_id,
      "TunnelSecret" = cloudflare_argo_tunnel.tunnel.secret,
      "TunnelID"     = cloudflare_argo_tunnel.tunnel.id,
      "TunnelName"   = cloudflare_argo_tunnel.tunnel.name
    })
  }
}

resource "random_string" "tunnel-secret" {
  length  = 32
  special = true
}

resource "cloudflare_argo_tunnel" "tunnel" {
  account_id = var.cloudflare_account_id
  name       = "argo-tunnel-ingress.${var.url}"
  secret     = base64encode(random_string.tunnel-secret.result)
}

resource "cloudflare_load_balancer" "loadbalancer" {
  zone_id = var.cloudflare_zone_id
  name = local.ingress_url
  fallback_pool_id = cloudflare_load_balancer_pool.ingress.id
  default_pool_ids = [cloudflare_load_balancer_pool.ingress.id]
  description = "Load Balancer for Staging"
  proxied = true
  steering_policy = "off"
  session_affinity = "cookie"
}

resource "cloudflare_load_balancer_pool" "ingress" {
  name = local.lb_pool_name
  description = "Load balancer pool for staging"
  enabled = true
  minimum_origins = 1
  monitor = cloudflare_load_balancer_monitor.monitor.id
  notification_email = "me@example.com"
  origins {
    address = "${cloudflare_argo_tunnel.tunnel.id}.cfargotunnel.com"
    name = "tunnel"
  }
}
```
and this is the ingress-nginx sidecar definition:
```
  extraContainers:
    - name: tunnel
      image: docker.io/cloudflare/cloudflared:2021.2.4
      imagePullPolicy: Always
      command: ["cloudflared"]
      args:
        - tunnel
        - run
        - --url=http://127.0.0.1:80
        - --cred-file=/etc/cloudflared/credentials.json
        - ${tunnel_id}
      resources:
        requests:
          cpu: 100m
          memory: 100Mi
      volumeMounts:
        - mountPath: /etc/cloudflared
          name: tunnel-secret
          readOnly: true
  extraVolumes:
    - name: tunnel-secret
      secret:
        secretName: ${secret_name}
```
It's still throwing lots of errors though:
2021-03-01T13:50:08Z ERR Register tunnel error from server side error="cannot replace existing connection" connIndex=2
2021-03-01T13:50:09Z ERR Register tunnel error from server side error="failed to register connection with backend" connIndex=3
2021-03-01T13:50:10Z ERR Connection terminated error="cannot replace existing connection" connIndex=1
:woman_shrugging:

**Comment by nin-ja-b** [comment]
So, in case anyone finds this. I got it fixed. I had 3 ingress-nginx replicas running that all used the same tunnel and failed. So now I got two different ingress-nginx deployments with 2 different tunnels and it magically works ;)
